### ISSUE TITLE
Move CSV Import Functionality to Core

### ISSUE THREAD

**Issue by SDavisMedia** [easydigitaldownloads/easy-digital-downloads#3620]
After discussing it in the team meeting and chatting with @pippinsplugins, this functionality is going to shift from extension to EDD core. This is where it begins. :+1:

**Comment by pippinsplugins** [comment]
This should include the following:

- payment import
- product import
- customer import (maybe)

Along with this introduction, I think we should make export and import from EDD site A to EDD site B seamless. If export a purchase history file from one site, I should be able to directly import it into another site without hassle.

**Comment by pippinsplugins** [comment]
I'm going to start working on this.

**Comment by pippinsplugins** [comment]
This is now up to date with the latest `release/2.6`

**Comment by pippinsplugins** [comment]
I have the base product import working well. Now just need to go through and handle all of the field mapping.

**Comment by pippinsplugins** [comment]
Ok this could use some initial testing for the product import section.

For testing, please start with a product export from EDD and then import that. Once that's working well, we can move to other file structures too.

Some notes:

1. UI is still rough.
2. Only product import is functional.
3. Automatically mapping CSV columns to product fields is not yet supported so for now all fields have to be mapped manually.
4. Category and tag fields are not working yet and am not sure why.
5. Featured image doesn't work yet.
6. Custom fields don't work yet.

**Comment by SDavisMedia** [comment]
My initial test with two sites on `issue/3620` is being met with this on a white screen during the import:

![screen shot 2016-05-10 at 9 07 21 pm](https://cloud.githubusercontent.com/assets/2359131/15169719/4092ebcc-16f3-11e6-9080-5648c0db4f62.png)

All I'm doing is exporting CSV from EDD core, uploading the CSV file for import, and when I click the Import button, that's what I get.

**Comment by SDavisMedia** [comment]
Scratch that... it worked from my live site to my local. No clue why it did that from local to live.

Still testing... more to follow.

**Comment by SDavisMedia** [comment]
Ignoring the fields you mentioned above, the import worked perfectly! No issues at all and all the product meta is correct. 🎉 

Would be cool to have a link to the downloads table here: http://glui.me/?i=gehkxgpydo4xiqs/2016-05-10_at_9.16_PM.png/

**Comment by SDavisMedia** [comment]
Actually, it looks like Author ->  Product Author didn't work.

**Comment by pippinsplugins** [comment]
@SDavisMedia I think that white screen might have been due to a JS file being cached.

**Comment by pippinsplugins** [comment]
`Import complete` message now includes a link to the relevant list table.

**Comment by michaelbeil** [comment]
It looks like we should add the `Needs Testing` label here and also create a PR for `issue/3620` against `release/2.6`. 

https://github.com/easydigitaldownloads/easy-digital-downloads/compare/release/2.6...issue/3620

**Comment by pippinsplugins** [comment]
It does need testing but it's also not nearly done, thus why I hadn't added it :)

**Comment by pippinsplugins** [comment]
Featured image import now works. Images can be specified by:

- Filename.ext - File must exists somewhere inside the `wp-content/uploads/` directory for this to work. Can be in any sub folder.
- http://site.com/wp-content/uploads/2016/05/filename.ext - URL pointing to file in uploads folder.
- http://some-external-site.com/file.ext - An external image file that will be downloaded to the WP media library

**Comment by pippinsplugins** [comment]
The product import is ready for another round of testing!

**Comment by SDavisMedia** [comment]
Getting the same behavior as this again: https://github.com/easydigitaldownloads/easy-digital-downloads/issues/3620#issuecomment-218356870

I'm exporting from a local MAMP site to a _brand new_ local MAMP site. Literally nothing has been done to the new site previously.

**Comment by pippinsplugins** [comment]
@SDavisMedia Force reload the page. It's a JS caching issue (as far as I can tell).

**Comment by SDavisMedia** [comment]
@pippinsplugins JS cached where? It could only be browser and that's been cleared. Trying to import to a site that was just created about an hour ago.

Could the data itself cause this? I'm attaching my compressed CSV file. 
[edd-export-downloads-05-25-2016.csv.zip](https://github.com/easydigitaldownloads/easy-digital-downloads/files/282560/edd-export-downloads-05-25-2016.csv.zip)

**Comment by pippinsplugins** [comment]
Do you have `SCRIPT_DEBUG` on?

The CSV file shouldn't cause that. It's happening because the browser is loading the uploaded file instead letting the JS read it. So far I've only seen that when the browser is loading a cached version of the JS file.

**Comment by SDavisMedia** [comment]
It was definitely because of `SCRIPT_DEBUG` not being true. No clue why there'd be a cached file from this new site.

**Comment by pippinsplugins** [comment]
Browsers cache asset files even without caching plugins active.

**Comment by SDavisMedia** [comment]
Got it.

For some reason, I cannot get the system to import more than 19 products. @michaelbeil experienced the same issue but he was using my CSV file. I've deleted some downloads and added new ones on the export site, keeping the total over 19. Tried a new export/import and it stops at 19 again.

Here's a sample CSV with 20 downloads: 
[20-items.csv.zip](https://github.com/easydigitaldownloads/easy-digital-downloads/files/282907/20-items.csv.zip)

**Comment by michaelbeil** [comment]
With a new set of 20 downloads in a CSV, only 19 were imported.

**Comment by pippinsplugins** [comment]
@SDavisMedia @michaelbeil could you try again please?

**Comment by pippinsplugins** [comment]
Validation of empty / non-CSV files has been restored.

**Comment by pippinsplugins** [comment]
Payment importing now works!

**Comment by SDavisMedia** [comment]
@pippinsplugins The issue with only importing 19 products is now fixed! :+1:

Not sure why no mapping is happening for me.

**Comment by pippinsplugins** [comment]
No mapping? Could you elabolate?

**Comment by SDavisMedia** [comment]
@pippinsplugins So far today, my fields are not automatically mapping when I import the CSV. For each test I've done, I'm setting them all manually. 

The exported CSV is coming from a site also running `issue/3620` but not the most recent changes from the last 2 hours. I'll pull down on the export site again to see if that makes a difference.

**Comment by pippinsplugins** [comment]
Ah, the magic mapping. That's not fully set up yet.

**Comment by SDavisMedia** [comment]
Okay all good [as far as we can be] with product exporting/importing at the moment.

---

**Payment and Customer exporting:**

With both of them, I found myself thinking I _had_ to choose a date range (Payment export) and make a Download section (Customer export) in order to get the CSVs. Obviously, that wouldn't work to get all data in one CSV for both. 

I only realized I could get all the data in both by just trying without using those filters. It worked, of course, but I can see people being stuck there. So that's a UI thing we may need to be more clear on.

**Payment import:**

Honestly had no idea that the Payments export and import functionality would be capable of creating my customer records as well. I thought I had to handle importing customers on my own. So that worked. However, I'll have to test again because secondary email addresses (from the new feature) on my customer records from the export site were considered new customers on the import site. New customer records were created for those email addresses. I guess that makes sense for the overall 2.6 release, though.

**Comment by pippinsplugins** [comment]
Exporting: for now, these will probably be opened as separate issues. I want to address each of them but probably not for beta 1 tomorrow. Will try for 2.6 still though. We need to first figure out _exactly_ which changes need to be made.

Import: yep! It will create customers along the way :) That won't work until we have everything merged into `release/2.6` since some of the code is still in separate issue branches.

**Comment by SDavisMedia** [comment]
@pippinsplugins Cool cool! Outside of that, everything was pretty smooth. :+1:

**Comment by pippinsplugins** [comment]
Great!

**Comment by michaelbeil** [comment]
All works well here after a `git pull` and a product import. I can now import over 19 products. We should be good to go for beta 1 tomorrow.

Export/Import UI referenced by @SDavisMedia above could use some improvements.

**Comment by pippinsplugins** [comment]
A few items I know we haven't finished yet:

- [ ] Help text for the payment importer after uploading the file
- [ ] Links to Import / Export tools page from the Reports > Exports page
- [ ] Variable price support for payment import
- [ ] Something funky is going on with payment keys. The keys appear to work fine but searching for a key after import fails
- [ ] Some UI love from @SDavisMedia

**Comment by pippinsplugins** [comment]
Also need:
- [ ] some form of error logging to report failed imports on a per-row basis

**Comment by SDavisMedia** [comment]
While testing Payments export/import, Products (raw) doesn't appear to be working properly. Some Products are being created upon import with names like `Product Name / Product Name / Product Name`. http://glui.me/?i=ugkcvhqymvgarls/2016-05-26_at_12.44_PM.png/

Here's my CSV if you need to take a look. 
[payments-may-26-1212.csv.zip](https://github.com/easydigitaldownloads/easy-digital-downloads/files/285215/payments-may-26-1212.csv.zip)

**Comment by michaelbeil** [comment]
Ran the same test with @SDavisMedia's CSSV and ran into the same issues.

**Comment by pippinsplugins** [comment]
Tweaks to make:

1. Add alternating row colors
2. Add some padding to select menus
3. Add help text (see #4641 )
4. Support CSVs without header rows (currently fails)
5. Show first row value next to each column for `CSV Column`
6. Show which fields are required
7. Put product field on left, CSV on right
8, Make product fields fixed. No need to select the field.
9. Test with RTL

**Comment by pippinsplugins** [comment]
Adjusted import UI:
![screen shot 2016-06-07 at 11 19 34 pm](https://cloud.githubusercontent.com/assets/1034109/15882642/529dbe22-2d06-11e6-854f-d397fd8d9ec1.png)

It's functional for both product and payment import.

**Comment by pippinsplugins** [comment]
In `issue/3620`

**Comment by pippinsplugins** [comment]
RTL works:
![screen shot 2016-06-08 at 12 04 38 am](https://cloud.githubusercontent.com/assets/1034109/15883300/9b4853b6-2d0c-11e6-93ff-45dc76c75158.png)

**Comment by michaelbeil** [comment]
I think it might be good to bold the column titles "Payment Field" and "CSV column".

**Comment by pippinsplugins** [comment]
Live data previews are in!
![screen shot 2016-06-08 at 11 42 29 am](https://cloud.githubusercontent.com/assets/1034109/15902564/1dcdc6d4-2d6e-11e6-9ffd-df1d26eed8ef.png)

Please test in `issue/3620` branch.

**Comment by SDavisMedia** [comment]
I'll give feedback as I go. First thing I notice on the Download import is this setting:

![2016-06-08 at 10 02 am](https://cloud.githubusercontent.com/assets/2359131/15903170/37dc9ffe-2d60-11e6-9e93-b8e960a2f8d9.png)

I definitely think the label needs tweaking. Something along the lines of "Check this box if blah blah." Blah blah being the least likely scenario for CSVs and header rows. For example, if you suspect that 80% of CSVs will have a header row, then "Check this box if your CSV file does not have a header row."

**Comment by cklosowski** [comment]
@SDavisMedia I can add a tooltip once this is ready and merged and I go through and add the helpful data to it.

**Comment by SDavisMedia** [comment]
Yup! And Pippin also stated that it will be checked by default as well. So all good there!

**Comment by michaelbeil** [comment]
Looking good here.

**Comment by SDavisMedia** [comment]
Decimal places on on Total Earnings preview data could be trimmed:

![screen shot 2016-06-08 at 10 11 35 am](https://cloud.githubusercontent.com/assets/2359131/15903407/78cbda92-2d61-11e6-8700-71455633dfae.png)

**Comment by SDavisMedia** [comment]
Just finished a product export and import and wow... what a smooth process. Super intuitive and the visual feedback from the Data Preview column is golden. That was amazingly easy. Two thumbs up.

Gonna do payments now.

**Comment by SDavisMedia** [comment]
Payment export/import was smooth as well. There was a little confusion on some of the mapping. I'm sure it'll improve with tool tips but just to keep record, I'll explain below. This was a CSV from EDD core.

---

Starting the list with **Currency Code** and me not having anything for it was discouraging right out of the gate. I lost a little confidence in my ability to map properly (and that the import would work) before I even got to make a successful match. Just something to think about for the inexperienced users.

I didn't map **Customer ID**. Wasn't sure what it should be.

Why aren't **IP Address** and **Mode** in the export?

**Payment Number** will need a strong tool tip since it can be a number of things and _sometimes_ match the Parent Payment ID itself.

No clue what **Subtotal** would be but as a user, it's not something I was comfortable ignoring considering it's extremely relevant to payment data. I had to ignore it, though, because there was nothing to map it to. Confidence killer.

I mapped "User" to **User ID** but I'm pretty sure that was wrong. Seems that should be in the CSV export file.

**Comment by pippinsplugins** [comment]
1. Currency code: added to payment export.

2. Customer ID. This one is tricky. Most of the time it should not be mapped, but sometimes it needs to be. See https://github.com/easydigitaldownloads/easy-digital-downloads/issues/4625. The main place it is needed is when going from EDD to EDD and customers are being imported that have multiple email addresses.

3. IP added to export.

4. Payment number: :+1:

5. Subtotal. This will be helped by a tool tip. It would be the purchase total before taxes usually.

6. User ID can be an email, username, or actual user Id.

**Comment by SDavisMedia** [comment]
Payment import affected by https://github.com/easydigitaldownloads/easy-digital-downloads/commit/195455b4d92c719114cdcc99efd82931fe45f75e

Haven't tested yet but I'd image the product import is affected as well. 

![2016-06-08 at 12 21 pm](https://cloud.githubusercontent.com/assets/2359131/15907881/d2b13af4-2d73-11e6-8a8d-f5592322982f.png)

**Comment by SDavisMedia** [comment]
That fixed it. I'll go back and verify product earnings data preview is still good shortly.

**Comment by SDavisMedia** [comment]
Payment import mapping is much better.

Not all of my payments imported successfully but I'm not sure why. I can't find several of the payment IDs in my history. I'm attaching the CSV file I imported.

[payments.csv.zip](https://github.com/easydigitaldownloads/easy-digital-downloads/files/305570/payments.csv.zip)

**Comment by SDavisMedia** [comment]
Maybe there's something up with my data. I remember this happening with my first round of testing some weeks back. Pretty sure @michaelbeil testing with my CSV and had the same issue. I'll need to test with new data.

**Comment by michaelbeil** [comment]
You are remembering correctly @SDavisMedia.

**Comment by pippinsplugins** [comment]
@SDavisMedia from that CSV, do you happen to know of at least one example that didn't import?

**Comment by SDavisMedia** [comment]
@pippinsplugins It's reeaaally hard to say because it's looking like the IDs aren't matching up either. 

What I mean is the very first product, ID 7359, did not import. The first one to import was row 215, ID 6271. But the CSV file's information for 6271 does not match ID 6271 data in my system after the import.

So not only is it hard to tell what didn't import, it's also hard to tell if my imports actually match my exports.

This data has been accumulating since 2014. Plenty of products and customers have been deleted. Not sure if that affects the export/import data in a negative way.

**Comment by pippinsplugins** [comment]
Note: IDs will not match before and after import.

What's the name of the product that didn't work?

**Comment by SDavisMedia** [comment]
Well I figured that was the case but then I wondered what's the point of mapping them? I guess for data relationship or something?

Anyway, again it's hard to say because some things are not matching. I simply can't tell what wasn't imported... I can only tell that the total number of records doesn't match up.

So let's look at the differences between one individual record that I've identified on both sites. The first product in my CSV is a transaction from June 6, 2016 (2016-06-06). I only have two transactions from that date in my export and they're pretty much identical purchases.

![screen shot 2016-06-08 at 10 23 07 pm](https://cloud.githubusercontent.com/assets/2359131/15919583/9f18f44e-2dc7-11e6-8c10-f32bf662a4a2.png)

Notice both transactions are for $60. The payment records are pretty much identical so we can just look at the most recent one.

![2016-06-08 at 10 11 pm](https://cloud.githubusercontent.com/assets/2359131/15919436/2ae2c768-2dc6-11e6-95fe-8a8037022a91.png)

My "Commissions" product is $100 with an _excluded_ 20% tax, making it $120. On that transaction, a discount code was used titled `discount` and it is for 50% off. That puts the transaction at $60.

---

Now moving over to that same record on my import site. Things I'm noticing:

1. The first record displaying in my export site (the one I too screenshots of above) is the _last_ record on my import site (notice the pagination, the 3rd record details, etc.):

![screen shot 2016-06-08 at 10 22 38 pm](https://cloud.githubusercontent.com/assets/2359131/15919588/a473a38a-2dc7-11e6-89e5-d966d42a5255.png)

Should it be reversed like that? My entire import site purchase history is backwards, showing 2014 records first.

2. Focusing on the total price, it's wrong displaying at $170 on the table. Clicking through to the actual record, it changes even more:

![screen shot 2016-06-08 at 10 26 53 pm](https://cloud.githubusercontent.com/assets/2359131/15919687/7811e3d2-2dc8-11e6-87f8-ed74753a35fc.png)

I can't really tell what the math is doing but I do know the import site is not accounting for tax settings or discounts that don't exist within its system.

---

Definitely having a hard time understanding what I'm seeing so I can't really answer these questions clearly.

Might be worth it to export from a site with 3 transactions, all different dates, and all with taxes and discounts. Then import to a site with no taxes or discounts and observe the price details and record ordering, if that makes sense.

**Comment by pippinsplugins** [comment]
@SDavisMedia There shouldn't be an option to map payment / product IDs. Can you show me a screenshot of the mapping?

1. Import order.

Looks like you didn't map the `Purchase Date`. If the purchase date is mapped, the order should be retained. I'm going to double check that we are importing them in the correct order--not immediately sure.

2. Taxes / discounts. I'm going to start testing!

**Comment by pippinsplugins** [comment]
Definitely found some problems.

The bottom four purchases were exported and then imported back in, so the bottom four should match the top four. The amounts are definitely wrong.
![screen shot 2016-06-09 at 9 43 44 am](https://cloud.githubusercontent.com/assets/1034109/15933802/c0beed60-2e26-11e6-9653-f9bac726b136.png)

**Comment by pippinsplugins** [comment]
@SDavisMedia you were correct about the export order. They were reversed.

**Comment by SDavisMedia** [comment]
My latest payment import attempt failed about a third of the way through. Here are the logged errors:

```
Jun 09, 20:55:23	PHP Notice: Indirect modification of overloaded property EDD_Payment::$address has no effect in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 342
Jun 09, 20:55:23	PHP Notice: Indirect modification of overloaded property EDD_Payment::$address has no effect in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 342
Jun 09, 20:55:23	PHP Notice: Indirect modification of overloaded property EDD_Payment::$address has no effect in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 342
Jun 09, 20:55:23	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:25	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:25	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:25	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:25	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:27	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:28	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:28	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:28	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:29	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:30	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:30	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:30	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:32	PHP Notice: Undefined offset: 1 in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 516
Jun 09, 20:55:32	PHP Notice: Undefined variable: customer_by_email in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 414
Jun 09, 20:55:32	PHP Notice: Trying to get property of non-object in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 427
Jun 09, 20:55:32	PHP Fatal error: Call to a member function create() on a non-object in /Applications/MAMP/htdocs/localhost/edd-testing/wp-content/plugins/easy-digital-downloads/includes/admin/import/class-batch-import-payments.php on line 441
```

**Comment by pippinsplugins** [comment]
Another notice I'm seeing:
```
PHP Notice: Undefined index: price_id in /Users/pippin/Sites/edd/wp-content/plugins/easy-digital-downloads/includes/admin/reporting/export/class-batch-export-payments.php on line 167
```

**Comment by pippinsplugins** [comment]
@SDavisMedia ready for another test!

**Comment by pippinsplugins** [comment]
Looks like the email metadata isn't getting saved completely during import.

Payment History:

![screen shot 2016-06-13 at 8 18 50 am](https://cloud.githubusercontent.com/assets/1034109/16008836/8453e4c2-313f-11e6-9a2c-27c81139b50f.png)

Dashboard widget and error log:

![screen shot 2016-06-13 at 8 18 55 am](https://cloud.githubusercontent.com/assets/1034109/16008838/88149b9c-313f-11e6-84ef-92f88fd0280a.png)

**Comment by pippinsplugins** [comment]
@easydigitaldownloads/core-devs could I get one more round of tests on this? Primarily want to verify my fix for the email issue above.

**Comment by mintplugins** [comment]
Just had a crash on import. 500 internal server error. Digging in to try and find the cause

**Comment by mintplugins** [comment]
Here is my error log (which I cleared prior to importing. These should all be related to the import.
[error-log-upon-import.txt](https://github.com/easydigitaldownloads/easy-digital-downloads/files/321846/error-log-upon-import.txt)

To confirm, I tested on branch issue/2.6 up to https://github.com/easydigitaldownloads/easy-digital-downloads/commit/27d96af946b3eb099ba1ca481731f52b43f970f5

**Comment by pippinsplugins** [comment]
@mintplugins could you test your same import again and let me know if the errors are gone?

**Comment by pippinsplugins** [comment]
See #4692 for future adjustments of column names.

**Comment by mintplugins** [comment]
Import worked flawlessly now! 👍

**Comment by pippinsplugins** [comment]
@mintplugins great!

I'm going to go ahead and close this as complete!
